packages/graph-explorer-proxy-server/node-server.js

```diff
@@ -9,74 +9,85 @@ const https = require("https");
 const fs = require("fs");
 const path = require("path");
 const { fromNodeProviderChain } = require("@aws-sdk/credential-providers");
+const aws4 = require("aws4");
 
 const getCredentials = async () => {
-  let credentials;
   try {
-    credentials = fromNodeProviderChain()
-    console.log("Master credentials available");
+    const credentialProvider = fromNodeProviderChain();
+    const results = await credentialProvider();
+    console.log(
+      "IAM credentials were found in provider chain and will be used to sign requests"
+    );
+    return results;
   } catch (e) {
-    console.error("No master credentials available", e);
+    console.error("No IAM credentials found in provider chain", e);
   }
-  return credentials;
 };
 
-dotenv.config({ path: "../graph-explorer/.env" });
-
-(async () => {
+async function getIAMHeaders(options) {
   let creds = await getCredentials();
-  let requestSig;
-
-  app.use(cors());
-
-  app.use(process.env.GRAPH_EXP_ENV_ROOT_FOLDER, express.static(path.join(__dirname, "../graph-explorer/dist")));
+  if (creds === undefined) {
+    throw new Error(
+      "IAM is enabled but credentials cannot be found on the credential provider chain."
+    );
+  }
+  const headers = aws4.sign(options, {
+    accessKeyId: creds.accessKeyId,
+    secretAccessKey: creds.secretAccessKey,
+    sessionToken: creds.sessionToken,
+  }).headers;
 
-  const delay = (ms) => new Promise((resolve) => setTimeout(() => resolve(), ms));
+  return headers;
+}
 
-  async function retryFetch (
-    url,
-    retries = 1,
-    retryDelay = 10000,
-    req,
-    language
-  ) {
+dotenv.config({ path: "../graph-explorer/.env" });
 
-    let reqObjects;
+(async () => {
+  app.use(cors());
 
-    if (creds === undefined) {
-      console.error("Credentials undefined. Trying refresh.");
-      let creds = await getCredentials();
-      if (creds === undefined) {
-        throw new Error("Credentials still undefined. Check that the environment has an appropriate IAM role that trusts it and that it has sufficient read permissions to connect to Neptune.")
-      }
+  app.use(
+    process.env.GRAPH_EXP_ENV_ROOT_FOLDER,
+    express.static(path.join(__dirname, "../graph-explorer/dist"))
+  );
+
+  const delay = (ms) =>
+    new Promise((resolve) => setTimeout(() => resolve(), ms));
+
+  async function retryFetch(url, headers, retries = 1, retryDelay = 10000) {
+    if (headers["aws-neptune-region"]) {
+      await getIAMHeaders({
+        host: url.hostname,
+        port: url.port,
+        path: url.pathname + url.search,
+        service: "neptune-db",
+        region: headers["aws-neptune-region"],
+      }).then((data) => {
+        headers = data;
```

Comment:
If I understand correctly, we are replacing the `headers` passed to the function with the IAM headers here. 
Q. Is there a risk that if any headers were passed in the request, they would be removed and if so, should we append the IAM headers to `headers` instead of replacing?

Reply:
done.